### ISSUE TITLE
Potential Problems in dp_rank Calculation and Gradient Synchronization Efficiency

### ISSUE THREAD

**Issue by Hannibal046** [facebookresearch/lingua#66]
Hi Team,

First off, thank you for maintaining such a clean and well-structured codebase! It's been a pleasure exploring it. While doing so, I came across two potential issues I would like to bring to your attention.

#### 1. Incorrect dp_rank calculation when dp_shard is used
The calculation of dp_rank in the following snippet may not be handling sharding correctly when dp_shard is involved:

https://github.com/facebookresearch/lingua/blob/0638b7b2a50684863b39e4a235cb3b627e9801be/apps/main/train.py#L243

However, I believe it should be updated to:
```python
dp_rank = dp_rank * world_mesh["dp_shard"].size() + world_mesh["dp_shard"].get_local_rank()
```

Failing to do so seems to cause issues similar to the one reported [here](https://github.com/facebookresearch/lingua/issues/58), as the parameter dp_rank will not properly account for the local rank within the shard.

---

#### 2. Optimization: Reduce Communication Overhead with gradient_accumulation_steps > 1
When using gradient accumulation (gradient_accumulation_steps > 1), communication overhead can be reduced by restricting gradient synchronization to the last micro-batch within an accumulation step. This can be achieved by adding:
```python
model.set_requires_gradient_sync(train_state.acc_step == 0)
```
just before the `loss.backward()` call.

Rationale: Since gradients only need to be synchronized after the full accumulation range is completed, synchronizing in all steps unnecessarily increases communication cost.

---

Thank you for taking the time to review these suggestions! I hope this feedback is helpful. Please let me know if additional clarification or testing is required on my part.

**Comment by Hannibal046** [comment]
Same here:
https://github.com/facebookresearch/lingua/blob/0638b7b2a50684863b39e4a235cb3b627e9801be/lingua/checkpoint.py#L196

**Comment by mathuvu** [comment]
Thanks for your help! We have added this into PR #70, which has been merged in main.
